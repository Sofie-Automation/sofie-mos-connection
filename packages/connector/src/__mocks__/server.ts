/* eslint-disable @typescript-eslint/unbound-method */
import { EventEmitter } from 'events'
import { Server } from 'net'
import { SocketMock } from './socket'

import * as iconv from 'iconv-lite'
iconv.encodingExists('utf16-be')

// Mock the Server class in 'net':
const setTimeoutOrg = setTimeout

const instances: Array<IServerMock> = []

export class ServerMock extends EventEmitter implements Server {
	// Mock implementation:

	maxConnections = 0
	connections = 0
	listening = false
	listenToPort = 0

	private _responses: Array<(data: any) => string | Buffer> = []

	constructor() {
		super()

		// @ts-expect-error this is comparable with ISocketMock
		instances.push(this)

		this.listen = jest.fn(this.listen)
		this.close = jest.fn(this.close)
		this.address = jest.fn(this.address)
		this.getConnections = jest.fn(this.getConnections)
		this.ref = jest.fn(this.ref)
		this.unref = jest.fn(this.unref)
	}
	static mockClear(): void {
		instances.splice(0, 9999)
	}
	static get instances(): IServerMock[] {
		return instances
	}

	// @ts-expect-error mock hack
	listen(port: number): this {
		this.listenToPort = port
		setTimeoutOrg(() => {
			this.listening = true
			this.emit('listening', {})
		}, 1)
		return this
	}
	close(callback?: () => void): this {
		if (callback) callback()
		setTimeoutOrg(() => {
			this.emit('close')
		}, 1)
		return this
	}
	address(): { port: number; family: string; address: string } {
		return { port: 0, family: 'string', address: 'string' }
	}
	getConnections(cb: (error: Error | null, count: number) => void): this {
		cb(null, 0)
		return this
	}
	ref(): this {
		return this
	}
	unref(): this {
		return this
	}

	// Mock methods:
	mockNewConnection(): SocketMock {
		// "Someone has connected"
		const socket = new SocketMock()
		socket.on('data', (d) => {
			this.emit('data', d)
		})

		this.emit('connection', socket)

		return socket
	}

	mockSentMessage(data: string, _encoding: unknown): void {
		const cb = this._responses.shift()
		if (cb) {
			// send reply:
			let msg

			setTimeoutOrg(() => {
				if (typeof cb === 'string') {
					msg = cb
				} else {
					msg = cb(data)
				}

				this.mockReceiveMessage(msg)
			}, 1)
		}
	}
	mockReceiveMessage(msg: string | Buffer): void {
		this.emit('data', msg)
	}
	mockAddReply(cb: (data: any) => string | Buffer): void {
		this._responses.push(cb)
	}
	mockClear(): void {
		this._responses.splice(0, 9999)
	}
	decode(data: Buffer): string {
		return iconv.decode(data, 'utf16-be')
	}
	encode(str: string): Buffer {
		return iconv.encode(str, 'utf16-be')
	}
}

export type IServerMock = jest.Mocked<ServerMock>
