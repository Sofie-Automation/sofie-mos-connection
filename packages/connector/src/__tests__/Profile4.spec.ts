import {
	checkMessageSnapshot,
	checkReplyToServer,
	clearMocks,
	decode,
	doBeforeAll,
	encode,
	fakeIncomingMessage,
	fixSnapshot,
	getMessageId,
	getMosConnection,
	getMosDevice,
	getReplyToServer,
	getXMLReply,
	mosTypes,
	setupMocks,
} from './lib'
import { MosConnection, MosDevice, IMOSROAck, IMOSROFullStory, IMOSRunningOrder } from '..'
import { SocketMock } from '../__mocks__/socket'
import { ServerMock } from '../__mocks__/server'
import { xmlData, xmlApiData } from '../__mocks__/testData'

/* eslint-disable @typescript-eslint/no-unused-vars */
// @ts-ignore imports are unused
import { Socket } from 'net'
/* eslint-enable @typescript-eslint/no-unused-vars */

beforeAll(() => {
	setupMocks()
})
beforeEach(() => {
	clearMocks()
})

describe('Profile 4', () => {
	let mosDevice: MosDevice
	let mosConnection: MosConnection
	let socketMockLower: SocketMock
	let socketMockUpper: SocketMock

	let serverSocketMockLower: SocketMock
	let serverSocketMockUpper: SocketMock
	let serverSocketMockQuery: SocketMock

	let onROStory: jest.Mock<any, any>

	beforeAll(async () => {
		SocketMock.mockClear()
		ServerMock.mockClear()

		mosConnection = await getMosConnection(
			{
				'0': true,
				'1': true,
				'2': true,
				'4': true,
			},
			false
		)
		mosDevice = await getMosDevice(mosConnection)

		const roAckReply = async () => {
			const ack: IMOSROAck = {
				ID: mosTypes.mosString128.create('runningOrderId'),
				Status: mosTypes.mosString128.create('OK'),
				Stories: [],
			}
			return ack
		}
		// Profile 4:
		onROStory = jest.fn(roAckReply)

		mosDevice.onRunningOrderStory(async (story: IMOSROFullStory): Promise<IMOSROAck> => {
			return onROStory(story)
		})
		const b = doBeforeAll()
		socketMockLower = b.socketMockLower
		socketMockUpper = b.socketMockUpper
		serverSocketMockLower = b.serverSocketMockLower
		serverSocketMockUpper = b.serverSocketMockUpper
		serverSocketMockQuery = b.serverSocketMockQuery

		mosConnection.checkProfileValidness()
		mosDevice.checkProfileValidness()
	})
	beforeEach(() => {
		onROStory.mockClear()

		serverSocketMockLower.mockClear()
		serverSocketMockUpper.mockClear()
		serverSocketMockQuery.mockClear()
	})
	afterAll(async () => {
		await mosDevice.dispose()
		await mosConnection.dispose()
	})
	test('init', async () => {
		expect(mosDevice).toBeTruthy()
		expect(socketMockLower).toBeTruthy()
		expect(socketMockUpper).toBeTruthy()
		mosDevice.checkProfileValidness()
	})
	test('onRunningOrderStory', async () => {
		// Fake incoming message on socket:

		const messageId = await fakeIncomingMessage(serverSocketMockLower, xmlData.roStorySend)
		expect(onROStory).toHaveBeenCalledTimes(1)

		const o = Object.assign({}, xmlApiData.roStorySend)
		// @ts-ignore optional property
		delete o.Body
		expect(onROStory.mock.calls[0][0]).toMatchObject(o)

		expect(fixSnapshot(onROStory.mock.calls)).toMatchSnapshot()
		xmlApiData.roStorySend.Body.forEach((testItem, key) => {
			let item: any
			try {
				item = onROStory.mock.calls[0][0].Body[key]
				if (!testItem.Content) delete testItem.Content

				expect(item).toMatchObject(testItem)
			} catch (e) {
				console.error(key)
				console.error('item', item)
				console.error('testItem', testItem)
				throw e
			}
		})
		await checkReplyToServer(serverSocketMockLower, messageId, '<roAck>')
	})
	test('onRequestAllRunningOrders', async () => {
		const onRoReqAll = jest.fn(async (): Promise<IMOSRunningOrder[]> => {
			return [
				{
					ID: mosTypes.mosString128.create('96857485'),
					Slug: mosTypes.mosString128.create('the slug'),
					DefaultChannel: mosTypes.mosString128.create('14'),
					EditorialStart: mosTypes.mosTime.create('2005-07-01T13:15:00Z'),
					EditorialDuration: mosTypes.mosDuration.create('1:23:45'),
					Trigger: mosTypes.mosString128.create('A'),
					Stories: [],
				},
				{
					ID: mosTypes.mosString128.create('96857486'),
					Slug: mosTypes.mosString128.create('the slug2'),
					DefaultChannel: mosTypes.mosString128.create('19'),
					Stories: [],
				},
			]
		})
		mosDevice.onRequestAllRunningOrders(onRoReqAll)

		// Fake incoming message on socket:
		const messageId = await fakeIncomingMessage(serverSocketMockLower, xmlData.roReqAll)
		expect(onRoReqAll).toHaveBeenCalledTimes(1)

		const reply = await getReplyToServer(serverSocketMockLower, messageId)
		expect(reply.mos.roListAll).toMatchObject({
			ro: [
				{
					roID: { _text: 96857485 },
					roSlug: { _text: 'the slug' },
					roChannel: { _text: 14 },
					roEdStart: { _text: '2005-07-01T13:15:00,000Z' },
					roEdDur: { _text: '1:23:45' },
					roTrigger: { _text: 'A' },
				},
				{
					roID: { _text: 96857486 },
				},
			],
		})
	})
	test('getAllRunningOrders', async () => {
		// Prepare server response
		const mockReply = jest.fn((data) => {
			const str = decode(data)
			const messageID = getMessageId(str)
			return encode(getXMLReply(messageID, xmlData.roListAll))
		})
		socketMockUpper.mockAddReply(mockReply)
		const returnedListAll = await mosDevice.sendRequestAllRunningOrders()
		await socketMockUpper.mockWaitForSentMessages()
		expect(mockReply).toHaveBeenCalledTimes(1)
		const msg = decode(mockReply.mock.calls[0][0])
		expect(msg).toMatch(/<roReqAll/)
		checkMessageSnapshot(msg)

		expect(returnedListAll).toHaveLength(2)
		expect(returnedListAll[0]).toMatchObject(xmlApiData.roListAll[0])
		expect(returnedListAll[1]).toMatchObject(xmlApiData.roListAll[1])
		expect(returnedListAll).toMatchSnapshot()
	})
	// onRequestAllRunningOrders
	// sendRunningOrderStory
})
