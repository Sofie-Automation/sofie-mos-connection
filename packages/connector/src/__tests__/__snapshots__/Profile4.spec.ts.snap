// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Profile 4 > getAllRunningOrders 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>xx</messageID>
  <roReqAll/>
</mos>"
`;

exports[`Profile 4 > getAllRunningOrders 2`] = `
[
  {
    "EditorialDuration": {
      "_mosDuration": 1800,
    },
    "EditorialStart": {
      "_mosTime": 1247331600000,
      "_timezone": "Z",
      "_timezoneOffset": 0,
    },
    "ID": {
      "_mosString128": "5PM",
    },
    "MosExternalMetaData": [
      {
        "MosPayload": {
          "Approved": "0",
          "Creator": "SHOLMES",
          "ModBy": "LJOHNSTON",
          "Owner": "SHOLMES",
          "TextTime": "278",
          "mediaTime": "0",
        },
        "MosSchema": "https://ncsA4.com/mos/supported_schemas/NCSAXML2.08",
        "MosScope": "PLAYLIST",
      },
    ],
    "Slug": {
      "_mosString128": "5PM Rundown",
    },
    "Trigger": {
      "_mosString128": "MANUAL",
    },
  },
  {
    "EditorialDuration": {
      "_mosDuration": 1800,
    },
    "EditorialStart": {
      "_mosTime": 1247162400000,
      "_timezone": "Z",
      "_timezoneOffset": 0,
    },
    "ID": {
      "_mosString128": "6PM",
    },
    "MosExternalMetaData": [
      {
        "MosPayload": {
          "Approved": "1",
          "Creator": "SHOLMES",
          "ModBy": "BSMITH",
          "Owner": "SHOLMES",
          "TextTime": "350",
          "mediaTime": "0",
        },
        "MosSchema": "https://ncsA4.com/mos/supported_schemas/NCSAXML2.08",
        "MosScope": "PLAYLIST",
      },
    ],
    "Slug": {
      "_mosString128": "6PM Rundown",
    },
    "Trigger": {
      "_mosString128": "MANUAL",
    },
  },
]
`;

exports[`Profile 4 > onRunningOrderStory > <mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1633</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos> 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1633</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos>"
`;

exports[`Profile 4 > onRunningOrderStory 1`] = `
[
  [
    {
      "Body": [
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "ID": {
              "_mosString128": "2",
            },
            "MOSID": "METADATA.NRK.MOS",
            "MosExternalMetaData": [
              {
                "MosPayload": {
                  "nrk": {
                    "attributes": {
                      "changedBy": "N11580",
                      "changetime": "2014-10-06T13:24:32 +02:00",
                      "type": "video",
                    },
                    "description": "",
                    "hbbtv": {
                      "link": "",
                    },
                    "rights": {
                      "notes": "",
                      "owner": "NRK",
                      "text": "Green",
                    },
                    "title": "",
                  },
                },
                "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
                "MosScope": "PLAYLIST",
              },
            ],
            "ObjectID": {
              "_mosString128": "N11580_1412594672",
            },
            "ObjectSlug": {
              "_mosString128": "M:",
            },
            "Slug": {
              "_mosString128": "SAK BUSKERUD;SAK-14",
            },
            "mosAbstract": "METADATA",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "Channel": {
              "_mosString128": "CG1",
            },
            "ID": {
              "_mosString128": "3",
            },
            "MOSID": "chyron.techycami02.ndte.nrk.mos",
            "MosObjects": [
              {
                "Duration": 0,
                "ID": {
                  "_mosString128": "NYHETER\\00039287?version=1",
                },
                "MosAbstractStr": "",
                "MosExternalMetaData": [
                  {
                    "MosPayload": {
                      "ObjectDetails": {
                        "ServerID": "chyron.techycami02.ndte.n    --------------------------------------------------------    rk.mos",
                        "ServerURL": "https://160.68.33.159/CAMIO/Redirection/MOSRedirection.asmx",
                      },
                      "createdBy": "N12050",
                      "sAVeom": "00:00:05:00",
                      "sAVsom": "00:00:02:00",
                      "subtype": "lyric/data",
                      "subtypeid": "I:\\CAMIO\\NYHETER\\Templates\\Super\\00000001.lyr",
                    },
                    "MosSchema": "https://ncsA4.com/mos/supported_schemas/NCSAXML2.08",
                    "MosScope": "PLAYLIST",
                  },
                ],
                "MosItemEditorProgID": {
                  "_mosString128": "Chymox.AssetBrowser.1",
                },
                "Paths": [
                  {
                    "Description": "JPEG Thumbnail",
                    "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039287_v1_big.jpg",
                    "Type": "PROXY PATH",
                  },
                ],
                "Slug": {
                  "_mosString128": "01 ett navn 1:  2:",
                },
                "TimeBase": 0,
                "Type": "OTHER",
              },
            ],
            "ObjectID": {
              "_mosString128": "",
            },
            "Paths": [
              {
                "Description": "JPEG Thumbnail",
                "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039287_v1_big.jpg",
                "Type": "PROXY PATH",
              },
            ],
            "Slug": {
              "_mosString128": "01 ett navn 1:  2:",
            },
            "mosAbstract": "_00:00:02:00 | @M=Auto Timed | 01 ett navn | 1: | 2: | 3: | 4: | 00:00:05:00",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "Channel": {
              "_mosString128": "CG1",
            },
            "ID": {
              "_mosString128": "4",
            },
            "MOSID": "chyron.techycami02.ndte.nrk.mos",
            "MosObjects": [
              {
                "Duration": 0,
                "ID": {
                  "_mosString128": "NYHETER\\00039288?version=1",
                },
                "MosAbstractStr": "",
                "MosExternalMetaData": [
                  {
                    "MosPayload": {
                      "ObjectDetails": {
                        "ServerID": "chyron.techycami02.ndte.nrk.mos",
                        "ServerURL": "https://160.68.33.159/CAMIO/Redirection/MOSRedirection.asmx",
                      },
                      "createdBy": "N12050",
                      "sAVeom": "00:00:05:00",
                      "sAVsom": "00:00:02:00",
                      "subtype": "lyric/data",
                      "subtypeid": "I:\\CAMIO\\NYHETER\\Templates\\Super\\00000001.lyr",
                    },
                    "MosSchema": "https://ncsA4.com/mos/supported_schemas/NCSAXML2.08",
                    "MosScope": "PLAYLIST",
                  },
                ],
                "MosItemEditorProgID": {
                  "_mosString128": "Chymox.AssetBrowser.1",
                },
                "Paths": [
                  {
                    "Description": "JPEG Thumbnail",
                    "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039288_v1_big.jpg",
                    "Type": "PROXY PATH",
                  },
                ],
                "Slug": {
                  "_mosString128": "01 ett navn 1:  2:",
                },
                "TimeBase": 0,
                "Type": "OTHER",
              },
            ],
            "ObjectID": {
              "_mosString128": "",
            },
            "Paths": [
              {
                "Description": "JPEG Thumbnail",
                "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039288_v1_big.jpg",
                "Type": "PROXY PATH",
              },
            ],
            "Slug": {
              "_mosString128": "01 ett navn 1:  2:",
            },
            "mosAbstract": "_00:00:02:00 | @M=Auto Timed | 01 ett navn | 1: | 2: | 3: | 4: | 00:00:05:00",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "Channel": {
              "_mosString128": "CG1",
            },
            "ID": {
              "_mosString128": "5",
            },
            "MOSID": "chyron.techycami02.ndte.nrk.mos",
            "MosObjects": [
              {
                "Duration": 0,
                "ID": {
                  "_mosString128": "NYHETER\\00039289?version=1",
                },
                "MosAbstractStr": "",
                "MosExternalMetaData": [
                  {
                    "MosPayload": {
                      "ObjectDetails": {
                        "ServerID": "chyron.techycami02.ndte.nrk.mos",
                        "ServerURL": "https://160.68.33.159/CAMIO/Redirection/MOSRedirection.asmx",
                      },
                      "createdBy": "N12050",
                      "sAVeom": "00:00:05:00",
                      "sAVsom": "00:00:02:00",
                      "subtype": "lyric/data",
                      "subtypeid": "I:\\CAMIO\\NYHETER\\Templates\\Super\\00000001.lyr",
                    },
                    "MosSchema": "https://ncsA4.com/mos/supported_schemas/NCSAXML2.08",
                    "MosScope": "PLAYLIST",
                  },
                ],
                "MosItemEditorProgID": {
                  "_mosString128": "Chymox.AssetBrowser.1",
                },
                "Paths": [
                  {
                    "Description": "JPEG Thumbnail",
                    "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039289_v1_big.jpg",
                    "Type": "PROXY PATH",
                  },
                ],
                "Slug": {
                  "_mosString128": "01 ett navn 1:  2:",
                },
                "TimeBase": 0,
                "Type": "OTHER",
              },
            ],
            "ObjectID": {
              "_mosString128": "",
            },
            "Paths": [
              {
                "Description": "JPEG Thumbnail",
                "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039289_v1_big.jpg",
                "Type": "PROXY PATH",
              },
            ],
            "Slug": {
              "_mosString128": "01 ett navn 1:  2:",
            },
            "mosAbstract": "_00:00:02:00 | @M=Auto Timed | 01 ett navn | 1: | 2: | 3: | 4: | 00:00:05:00",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "Channel": {
              "_mosString128": "CG1",
            },
            "ID": {
              "_mosString128": "6",
            },
            "MOSID": "chyron.techycami02.ndte.nrk.mos",
            "MosObjects": [
              {
                "Duration": 0,
                "ID": {
                  "_mosString128": "NYHETER\\00039290?version=1",
                },
                "MosAbstractStr": "",
                "MosExternalMetaData": [
                  {
                    "MosPayload": {
                      "ObjectDetails": {
                        "ServerID": "chyron.techycami02.ndte.nrk.mos",
                        "ServerURL": "https://160.68.33.159/CAMIO/Redirection/MOSRedirection.asmx",
                      },
                      "createdBy": "N12050",
                      "sAVeom": "00:00:05:00",
                      "sAVsom": "00:00:02:00",
                      "subtype": "lyric/data",
                      "subtypeid": "I:\\CAMIO\\NYHETER\\Templates\\Super\\00000001.lyr",
                    },
                    "MosSchema": "https://ncsA4.com/mos/supported_schemas/NCSAXML2.08",
                    "MosScope": "PLAYLIST",
                  },
                ],
                "MosItemEditorProgID": {
                  "_mosString128": "Chymox.AssetBrowser.1",
                },
                "Paths": [
                  {
                    "Description": "JPEG Thumbnail",
                    "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039290_v1_big.jpg",
                    "Type": "PROXY PATH",
                  },
                ],
                "Slug": {
                  "_mosString128": "01 ett navn 1:  2:",
                },
                "TimeBase": 0,
                "Type": "OTHER",
              },
            ],
            "ObjectID": {
              "_mosString128": "",
            },
            "Paths": [
              {
                "Description": "JPEG Thumbnail",
                "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039290_v1_big.jpg",
                "Type": "PROXY PATH",
              },
            ],
            "Slug": {
              "_mosString128": "01 ett navn 1:  2:",
            },
            "mosAbstract": "_00:00:02:00 | @M=Auto Timed | 01 ett navn | 1: | 2: | 3: | 4: | 00:00:05:00",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "Channel": {
              "_mosString128": "CG1",
            },
            "ID": {
              "_mosString128": "7",
            },
            "MOSID": "chyron.techycami02.ndte.nrk.mos",
            "MosObjects": [
              {
                "Duration": 0,
                "ID": {
                  "_mosString128": "NYHETER\\00039291?version=1",
                },
                "MosAbstractStr": "",
                "MosExternalMetaData": [
                  {
                    "MosPayload": {
                      "ObjectDetails": {
                        "ServerID": "chyron.techycami02.ndte.nrk.mos",
                        "ServerURL": "https://160.68.33.159/CAMIO/Redirection/MOSRedirection.asmx",
                      },
                      "createdBy": "N12050",
                      "sAVeom": "00:00:05:00",
                      "sAVsom": "00:00:02:00",
                      "subtype": "lyric/data",
                      "subtypeid": "I:\\CAMIO\\NYHETER\\Templates\\Super\\00000024.lyr",
                    },
                    "MosSchema": "https://ncsA4.com/mos/supported_schemas/NCSAXML2.08",
                    "MosScope": "PLAYLIST",
                  },
                ],
                "MosItemEditorProgID": {
                  "_mosString128": "Chymox.AssetBrowser.1",
                },
                "Paths": [
                  {
                    "Description": "JPEG Thumbnail",
                    "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039291_v1_big.jpg",
                    "Type": "PROXY PATH",
                  },
                ],
                "Slug": {
                  "_mosString128": "24 foto/red 1:Foto og redigering:  2:",
                },
                "TimeBase": 0,
                "Type": "OTHER",
              },
            ],
            "ObjectID": {
              "_mosString128": "",
            },
            "Paths": [
              {
                "Description": "JPEG Thumbnail",
                "Target": "https://160.68.33.159/thumbs/NYHETER/39000/Objects_NYHETER_00039291_v1_big.jpg",
                "Type": "PROXY PATH",
              },
            ],
            "Slug": {
              "_mosString128": "24 foto/red 1:Foto og redigering:  2:",
            },
            "mosAbstract": "_00:00:02:00 | @M=Auto Timed | 24 foto/red | 1:Foto og redigering: | 2: | 3: | 4: | 00:00:05:00",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
        {
          "Content": {
            "ID": {
              "_mosString128": "8",
            },
            "MOSID": "mosart.morten.mos",
            "ObjectID": {
              "_mosString128": "STORYSTATUS",
            },
            "ObjectSlug": {
              "_mosString128": "Story status",
            },
            "Slug": {
              "_mosString128": "SAK BUSKERUD;SAK-20",
            },
            "mosAbstract": "TIDSMARKØR IKKE RØR",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "element",
          },
          "Type": "p",
          "itemType": "other",
        },
      ],
      "ID": {
        "_mosString128": "2012R2ENPS8VM;P_ENPSNEWS\\W\\R_696297DF-1568-4B36-B43B3B79514B40D4;1DAF0044-CA12-47BA-9F6CEFF33B3874FB",
      },
      "MosExternalMetaData": [
        {
          "MosPayload": {
            "Approved": "0",
            "Creator": "LINUXENPS",
            "ENPSItemType": "3",
            "MOSItemDurations": "0 0 0 0 0",
            "MOSItemEdDurations": "",
            "MOSObjSlugs": "M:   01 ett navn 1:  2:
   01 ett navn 1:  2:
   01 ett navn 1:  2:
   01 ett navn 1:  2:
   24 foto/red 1:Foto og redigering:  2:   Story status",
            "MOSSlugs": "SAK BUSKERUD;SAK-14
   01 ett navn 1:  2:
   01 ett navn 1:  2:
   01 ett navn 1:  2:
   01 ett navn 1:  2:
   24 foto/red 1:Foto og redigering:  2:   SAK BUSKERUD;SAK-20",
            "MediaTime": "0",
            "ModBy": "LINUXENPS",
            "ModTime": "20180227T004205Z",
            "Owner": "LINUXENPS",
            "SourceMediaTime": "0",
            "SourceTextTime": "0",
            "StoryProducer": "DKTE",
            "TextTime": "0",
            "mosartType": "FULL",
            "pubApproved": "0",
          },
          "MosSchema": "https://2012R2ENPS8VM:10505/schema/enps.dtd",
          "MosScope": "PLAYLIST",
        },
      ],
      "RunningOrderId": {
        "_mosString128": "2012R2ENPS8VM;P_ENPSNEWS\\W;696297DF-1568-4B36-B43B3B79514B40D4",
      },
      "Slug": {
        "_mosString128": "KRITIKK ETTER BRANN KONGSBERG;SAK",
      },
    },
  ],
]
`;

exports[`Profile 4 > onRunningOrderStory, string MosPayload > <mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1635</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos> 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1635</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos>"
`;

exports[`Profile 4 > onRunningOrderStory, string MosPayload 1`] = `
[
  [
    {
      "Body": [
        {
          "Content": {
            "ID": {
              "_mosString128": "2",
            },
            "MOSID": "METADATA.NRK.MOS",
            "MosExternalMetaData": [
              {
                "MosPayload": "THIS IS A STRING",
                "MosSchema": "https://MOSA4.com/mos/supported_schemas/MOSAXML2.08",
                "MosScope": "PLAYLIST",
              },
            ],
            "ObjectID": {
              "_mosString128": "N11580_1412594672",
            },
            "ObjectSlug": {
              "_mosString128": "M:",
            },
            "Slug": {
              "_mosString128": "SAK BUSKERUD;SAK-14",
            },
            "mosAbstract": "METADATA",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
        {
          "Content": {
            "$name": "p",
            "$type": "text",
            "text": "a text",
          },
          "Type": "p",
          "itemType": "other",
        },
      ],
      "ID": {
        "_mosString128": "2012R2ENPS8VM;P_ENPSNEWS\\W\\R_696297DF-1568-4B36-B43B3B79514B40D4;1DAF0044-CA12-47BA-9F6CEFF33B3874FB",
      },
      "MosExternalMetaData": [
        {
          "MosPayload": "THIS IS ANOTHER STRING",
          "MosSchema": "https://2012R2ENPS8VM:10505/schema/enps.dtd",
          "MosScope": "PLAYLIST",
        },
      ],
      "RunningOrderId": {
        "_mosString128": "2012R2ENPS8VM;P_ENPSNEWS\\W;696297DF-1568-4B36-B43B3B79514B40D4",
      },
      "Slug": {
        "_mosString128": "KRITIKK ETTER BRANN KONGSBERG;SAK",
      },
    },
  ],
]
`;

exports[`Profile 4 > onRunningOrderStory-single > <mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1634</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos> 1`] = `
"<mos>
  <ncsID>their.mos.id</ncsID>
  <mosID>our.mos.id</mosID>
  <messageID>1634</messageID>
  <roAck>
    <roID>runningOrderId</roID>
    <roStatus>OK</roStatus>
  </roAck>
</mos>"
`;

exports[`Profile 4 > onRunningOrderStory-single 1`] = `
[
  [
    {
      "Body": [
        {
          "Content": {
            "ID": {
              "_mosString128": "item0",
            },
            "MOSID": "mos0",
            "ObjectID": {
              "_mosString128": "object0",
            },
            "Trigger": "CHAINED",
          },
          "Type": "storyItem",
          "itemType": "storyItem",
        },
      ],
      "ID": {
        "_mosString128": "story0",
      },
      "RunningOrderId": {
        "_mosString128": "roID0",
      },
      "Slug": {
        "_mosString128": "My Story",
      },
    },
  ],
]
`;
